### PR TITLE
Re-enable Mouse Support

### PR DESCRIPTION
**Pull request by ergo720**
This PR adds a new KeyboardMouse device which allows cxbxr to receive input from both the keyboard and mouse simultaneously. Mouse bindings will use the "Axis X/Y/Z +/-" and "Click 0/1/2" control names in the input gui when bound. Existing keyboard profiles are retained and converted to use the new KeyboardMouse device.

### CODE REVIEW

File: src/common/IPCHybrid.hpp [hunk 1]
```diff
@@ -52,6 +52,6 @@ typedef enum class _IPC_UPDATE_KERNEL {
 	, CONFIG_INPUT_SYNC
 } IPC_UPDATE_KERNEL;
 
-void ipc_send_kernel_update(IPC_UPDATE_KERNEL command, const unsigned int value, const unsigned int hwnd);
```
Comment (RadWolfie) [inline src/common/IPCHybrid.hpp:55]: For what reason unsigned part is removed? It must be either signed or unsigned. Not omit either one out. Beside, it should be unsigned anyway.

Comment (ergo720) [reply src/common/IPCHybrid.hpp:55]: Because I need to pass the value `PORT_INVALID`, which is -1, and using an unsigned argument causes an unwanted  conversion. Also, the function is passing the value as `LPARAM` via `SendMessage`, which is a signed integer as well (it's `LONG_PTR`), so the passed unsigned value will be converted to signed anyway.

Comment (RadWolfie) [reply src/common/IPCHybrid.hpp:55]: Then simply replace it to `signed`.

**Comment by literalmente-game** [comment]
6afcd6a insta-crashes for me

**Comment by ergo720** [comment]
@literalmente-game  Yeah, I can confirm that as well. I have only tested by upgrading from an existing installation, which worked fine for me.

File: src/common/Settings.cpp [hunk 2]
```diff
@@ -893,4 +917,38 @@ void Settings::RemoveLegacyConfigs(unsigned int CurrentRevision)
 	default:
 		break;
 	}
+
+	if (CurrentRevision < 8) {
+		const std::string kb_str = "Keyboard";
+
+		for (unsigned port_num = 0; port_num < 4; ++port_num) {
+			std::string current_section = std::string(section_input_port) + std::to_string(port_num);
+			std::string device_name = m_si.GetValue(current_section.c_str(), sect_input_port.device);
```
Comment (ergo720) [inline src/common/Settings.cpp:926]: Problem found, it's crashing in this line

File: src/gui/resource/Cxbx.rc [hunk 3]
```diff
@@ -589,10 +603,10 @@ BEGIN
         MENUITEM "&Open Xbe...",                ID_FILE_OPEN_XBE,MFT_STRING,MFS_ENABLED
         MENUITEM "Open D&ashboard...\tF7",      ID_FILE_OPEN_DASHBOARD,MFT_STRING,MFS_ENABLED
         MENUITEM "&Close Xbe",                  ID_FILE_CLOSE_XBE,MFT_STRING,MFS_ENABLED
-        MENUITEM "", -1, MFT_SEPARATOR
```
Comment (RadWolfie) [inline src/gui/resource/Cxbx.rc:592]: I know Visual Studio tend to remove optional fields. However it is a requirement for Clang.

**Comment by ergo720** [comment]
Commit 01382d7 now allows users to lock the mouse cursor to the cxbxr rendering window when they press the F3 key. This is useful in windowed mode to avoid the cursor drifting to other windows while playing with the mouse. To release the cursor, simply press F3 again (it works like a toggle).

**Comment by literalmente-game** [comment]
From @LukeUsher on Discord 

> The problem with these toggles is they get forgotten when a new xbe is launched. In morrowind for example that constantly reboots itself
As a result, much of these toggle should store their state in emushared

**Comment by ergo720** [comment]
Commit dd24225 drops keyboard/mouse input when the mouse cursor is outside of the rendering window. This is again useful when playing with the mouse in windowed mode. This way, the game won't receive bogus input from the keyboard/mouse when the user is selecting/working on other windows.

File: src/gui/resource/Cxbx.rc [hunk 4]
```diff
@@ -27,6 +27,10 @@ LANGUAGE LANG_NEUTRAL, SUBLANG_NEUTRAL
 #ifdef APSTUDIO_INVOKED
 GUIDELINES DESIGNINFO
 BEGIN
+    IDD_INPUT_CFG, DIALOG
```
Comment (RadWolfie) [inline src/gui/resource/Cxbx.rc:30]: @ergo720 Did you add a new dialog then decided to revert the change?
Same thing for `IDD_INPUT_CFG AFX_DIALOG_LAYOUT` below.

File: src/gui/resource/ResCxbx.h [hunk 5]
```diff
@@ -304,6 +304,9 @@
 #define IDC_RUMBLE_TEST                 1302
 #define IDC_NETWORK_ADAPTER             1303
 #define IDC_LOG_POPUP_TESTCASE          1304
+#define IDC_INPUT_OPTIONS               1305
+#define IDC_MOUSE_RANGE                 1306
+#define IDC_WHEEL_RANGE                 1307
```
Comment (RadWolfie) [inline src/gui/resource/ResCxbx.h:309]: For some reason these additions did not update `_APS_NEXT_CONTROL_VALUE` value to 1308.

File: src/core/kernel/init/CxbxKrnl.h [hunk 6]
```diff
@@ -200,9 +200,11 @@ bool CxbxIsElevated();
 extern uint32_t CxbxKrnl_KernelThunkTable[379];
 
 extern bool g_bIsWine;
-
+extern bool g_bClipCursor;
 extern bool g_CxbxPrintUEM;
 extern ULONG g_CxbxFatalErrorCode;
+extern bool g_bIsTrackingMoLeave;
```
Comment (RadWolfie) [inline src/core/kernel/init/CxbxKrnl.h:206]: IMHO, these global should be confine in `DInputKeyboardMouse` files or in general input file.

**Comment by RadWolfie** [comment]
I have tested and confirmed mouse cursor is working as intended in-game. I'm not sure if Turok Evolution's menu selection with mouse is intended design or not. Since in-game does behave correctly, it may be just the game design.

I also have tested and confirmed mouse cursor does stay within render window in exclusive full screen, faux full screen, and windowed modes. Including multiple reboots.
👍

I am waiting on review remarks to be discuss or resolve them.

File: src/gui/resource/Cxbx.rc (hunk 4 above)
Comment (ergo720) [reply src/gui/resource/Cxbx.rc:30]: I didn't, I simply added the new options to the existing input gui. According to [here](https://stackoverflow.com/a/49444451), `DESIGNINFO` fields are only needed at design-time and don't change anything at runtime so this is a non-issue. BTW, `AFX_DIALOG_LAYOUT` (which also has a similar addition) is documented by Microsoft [here](https://docs.microsoft.com/en-us/cpp/mfc/dynamic-layout?view=vs-2019#to-set-the-dynamic-layout-programmatically-from-a-resource-file).

Comment (RadWolfie) [reply src/gui/resource/Cxbx.rc:30]: Thanks for clarification. Sometime I see resource editor randomly place something like this in the past, either at the top or at the bottom. It's hard to tell if it's intended design or not, or revert changes usually get left behind. Since they are usually empty like we see in here. 👍

**Review by LukeUsher — DISMISSED** [review]
I like it!

File: src/gui/DlgInputConfig.cpp [hunk 7]
```diff
@@ -183,16 +222,66 @@ INT_PTR CALLBACK DlgInputConfigProc(HWND hWndDlg, UINT uMsg, WPARAM wParam, LPAR
...
+	{
+		// Make sure we only allow decimal numbers and some special keys to delete characters
+		if (!((wParam >= '0' && wParam <= '9')
+			|| wParam == VK_CANCEL
+			|| wParam == VK_CLEAR
+			|| wParam == VK_DELETE
+			|| wParam == VK_BACK))
+		{
+			return FALSE;
+		}
```
Comment (RadWolfie) [inline src/gui/DlgInputConfig.cpp:277]: If you don't mind my asking, what was the purpose for this check?

File: src/gui/DlgInputConfig.cpp [hunk 8]
```diff
@@ -144,6 +178,11 @@ INT_PTR CALLBACK DlgInputConfigProc(HWND hWndDlg, UINT uMsg, WPARAM wParam, LPAR
 				{
 				case to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_DUKE): 
 				case to_underlying(XBOX_INPUT_DEVICE::MS_CONTROLLER_S): {
+					if (g_bHasOptChanges) {
+						UpdateInputOpt(hWndDlg);
+						g_InputDeviceManager.UpdateOpt(true);
+					}
```
Comment (RadWolfie) [inline src/gui/DlgInputConfig.cpp:184]: Do we need this code inside DeviceType switch statement? What about other controller types that are not standard controller? Since it is generic code. Perhaps we can check DeviceType's value itself outside of switch statement.

File: src/gui/DlgInputConfig.cpp (hunk 7 above)
Comment (ergo720) [reply src/gui/DlgInputConfig.cpp:277]: This check is critical. The text that the user inputs for the mouse range must be a valid decimal number because this text is later converted by `std::stol`. If it is not, the function throws an exception and because it will be unhanded, it will cause cxbxr to crash.

File: src/core/kernel/init/CxbxKrnl.h [hunk 9]
```diff
@@ -200,7 +200,7 @@ bool CxbxIsElevated();
 extern uint32_t CxbxKrnl_KernelThunkTable[379];
 
 extern bool g_bIsWine;
-
+extern bool g_bClipCursor;
```
Comment (RadWolfie) [inline src/core/kernel/init/CxbxKrnl.h:203]: TODO: Eventually move this extern global clip cursor once move some unrelate functions outside of Direct3D9 files. Such as `EmuMsgProc` function.

File: src/gui/DlgInputConfig.cpp (hunk 8 above)
Comment (ergo720) [reply src/gui/DlgInputConfig.cpp:184]: The purpose of this code is to synchronize the dinput mouse ranges with those that the user has chosen in the input gui. Without this, if the user has changed the ranges then, when they open the controller configuration, it will still use the old values, that is, those that were set when the input gui was first opened. This should be done regardless of the DeviceType, so that every xbox device under configuration will use the updated mouse ranges.

Comment (RadWolfie) [reply src/gui/DlgInputConfig.cpp:184]: I don't think you're getting the message. Couple lines down `DialogBoxParam` opens up duke configuration. Chihiro, Steel Battalion, dance pad, etc are not using standard controller. Which require a separate case with a different `DialogBoxParam` call to entirely different input configuration. That means it will require duplicate code you have here to the other controller type.

File: src/gui/DlgInputConfig.cpp (hunk 7 above)
Comment (RadWolfie) [reply src/gui/DlgInputConfig.cpp:277]: 👌
If I look at `EditControlSubclassProc` function and look at the rest of the code within this function. I would not know what exactly this function serve the purpose. Perhaps rename it to have mouse word in it or add a comment above it will help clarification.

Comment (LukeUsher) [reply src/gui/DlgInputConfig.cpp:277]: Why not try/catch around stol, and handle the exception gracefully? (Unless it's performance-critical)

Comment (ergo720) [reply src/gui/DlgInputConfig.cpp:277]: In my opinion preventing the problem from happening in the first place is cleaner than trying to handle it later. Also, what should happen if the user inputs a garbage (non-integer) value here? Open a MessageBox describing the problem? That's more bloated than simply forbidding anything but integers.

Comment (RadWolfie) [reply src/gui/DlgInputConfig.cpp:277]: Actually, that wouldn't work. This function is done identical to eeprom's UI which only accept certain characters. There should be input listener, which is what we have here, to accept certain characters than hexadecimal/paste random string. Without the input listener, users can insert anything they want. Which can cause invalid input from what we expect from users. I don't think there is a way around this foolproof design.

**Review by RadWolfie — APPROVED** [review]
All review remarks are resolve and had verified it is working as intended.
